desc/objectives/_neoclassical.py

```diff
@@ -0,0 +1,532 @@
+
+from desc.compute import get_profiles, get_transforms
+from desc.compute.utils import _compute as compute_fun
+from desc.grid import LinearGrid
+from desc.utils import Timer, setdefault
+
+from ..integrals import Bounce2D
+from ..integrals._quad_utils import (
+    automorphism_sin,
+    chebgauss2,
+    get_quadrature,
+    grad_automorphism_sin,
+)
+from ..integrals.basis import FourierChebyshevSeries
+from .objective_funs import _Objective, collect_docs
+from .utils import _parse_callable_target_bounds
+
+_bounce_overwrite = {
+    "deriv_mode": """
+    deriv_mode : {"auto", "fwd", "rev"}
+        Specify how to compute Jacobian matrix, either forward mode or reverse mode AD.
+        ``auto`` selects forward or reverse mode based on the size of the input and
+        output of the objective. Has no effect on ``self.grad`` or ``self.hess`` which
+        always use reverse mode and forward over reverse mode respectively.
+
+        Default is ``fwd``. If ``rev`` is chosen, then ``jac_chunk_size=1`` is chosen
+        by default. In ``rev`` mode, reducing the pitch angle parameter ``batch_size``
+        does not reduce memory, so it is recommended to retain the default for that.
+        """
+}
+
+
+class EffectiveRipple(_Objective):
+    """The effective ripple is a proxy for neoclassical transport.
+
+    The 3D geometry of the magnetic field in stellarators produces local magnetic
+    wells that lead to bad confinement properties with enhanced radial drift,
+    especially for trapped particles. Neoclassical (thermal) transport can become the
+    dominant transport channel in stellarators which are not optimized to reduce it.
+    The effective ripple is a proxy, measuring the effective modulation amplitude of the
+    magnetic field averaged along a magnetic surface, which can be used to optimize for
+    stellarators with improved confinement. It is targeted as a flux surface function.
+
+    References
+    ----------
+    https://doi.org/10.1063/1.873749.
+    Evaluation of 1/ν neoclassical transport in stellarators.
+    V. V. Nemov, S. V. Kasilov, W. Kernbichler, M. F. Heyn.
+    Phys. Plasmas 1 December 1999; 6 (12): 4622–4632.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        ``Equilibrium`` to be optimized.
+    grid : Grid
+        Optional, tensor-product grid in (ρ, θ, ζ) with uniformly spaced nodes
+        (θ, ζ) ∈ [0, 2π) × [0, 2π/NFP). Powers of two are preferable.
+    X : int
+        Grid resolution in poloidal direction for Clebsch coordinate grid.
+        Preferably power of 2.
+    Y : int
+        Grid resolution in toroidal direction for Clebsch coordinate grid.
+        Preferably power of 2.
+    Y_B : int
+        Desired resolution for algorithm to compute bounce points.
+        Default is double ``Y``. Something like 100 is usually sufficient.
+        Currently, this is the number of knots per toroidal transit over
+        to approximate |B| with cubic splines.
```

Comment:
is this still the case? I thought this one wasnt using splines?

Reply:
The splines are only used to compute the bounce points, and they are generated from a fourier series in (theta, zeta) (output of `rfft2`), which is much less expensive that to do it from FourierZernike.

Except for high NFP cases (e.g. heliotron), chebyshev approximation is significantly more efficient than splines. However, if I use chebyshev series for root finding with bounce points, we need to reverse mode differentiate through an eigenvalue solve. A naive reliance on a jnp.vectorize(orthax.chebroot)` would not be efficient for that. To use chebyshev approximation with AD, the algorithm I wrote out to avoid all but one of the eigenvalue solves per transit should be implemented. I stopped development work on algorithm; if I were to continue development on bounce integrals, I would be solving #1303 instead, i think probably with jax ffi stuff to hook in c code.

Reply:
https://github.com/PlasmaControl/DESC/pull/1119#discussion_r1868381236